feat(csharp): add examples for C# SDK

sobczal2 opened the pull request:
Closes #1891 

### Summary
This PR adds 4 usage examples, script for running them and updates CI to include that script.

### Details
All examples try to mirror existing rust examples. This was possible for:
- Basic
- GettingStarted
- MessageEnvelope
- MessageHeader

Skipped examples are:
- multi-tenant - `IggyProducer` type is not present in C# SDK, so I wasn't able to mirror it, this example is possible with some changes, but I think it is better to do it in a seperate PR, this one is already big
- new-sdk, sink-data-producer - I believe these are irrelevant, correct me if I'm wrong
- stream-builder - this can probably be implemented in the future, but there is no such API in C# SDK for now

Should `foreign/csharp/Iggy_Sample_Consumer` and `foreign/csharp/Iggy_Sample_Producer` be removed in this PR?

numinnex commented on examples/csharp/src/Basic/Iggy_SDK.Examples.Basic.Consumer/Program.cs:
```diff
@@ -0,0 +1,92 @@
+);
+
+var consumedBatches = 0;
+while (true)
```
I believe there is an api that uses `IAsyncEnumerable` and allows to use `async foreach`, you can use that to spin loop, rather than infinite loop. It does not make a difference in this case scenario, but I think it might be worth showing it

numinnex commented:
Left once comment, but LGTM! :smiley_cat:

lukaszzborek commented:
And yes, please remove old example projects from `foreign/csharp`

hubcio commented:
@sobczal2, for including the c# examples in this flow you just have to add it to examples suite in components.yml:
<img width="1114" height="365" alt="image" src="https://github.com/user-attachments/assets/d81833b8-2731-43f8-abd0-3f6524b013a2" />

sobczal2 commented:
Hopefully fixed CI issues and added examples to restore, build, lint. Refactored some steps to pass just sln/csproj directory instead of using `cd`, but I can revert it if you prefer it the other way

hubcio commented on .github/actions/csharp-dotnet/pre-merge/action.yml:
```diff
@@ -47,36 +47,34 @@ runs:
+        dotnet build foreign/csharp/Iggy_SDK.sln --no-restore /p:EnforceCodeStyleInBuild=true /p:TreatWarningsAsErrors=false
+        dotnet build examples/csharp/Iggy_SDK.Examples.sln --no-restore /p:EnforceCodeStyleInBuild=true /p:TreatWarningsAsErrors=false
 
         # TODO: make format check blocking (requires dotnet-format tool)
```
could you please fix this TODO? (remove || at the end and try to fix the lints)

sobczal2 replied:
This will invoke plenty of changes in the SDK itself. @lukaszzborek mentioned that he is working on a task there right now, it will be very difficult to merge, maybe we can fix it after that one? I can't even remove it just for examples, because examples sln references Iggy_SDK csproj, so dotnet format also checks it.

hubcio reviewed (commented):
overall great job! if possible, please fix these lints

hubcio commented:
@sobczal2 yeah, this is on my side, together with @spetz we introduced nasty bug with credentials for CI. working it on right now, will be merged with #2180 

after it's merged you can rebase and everything should work

hubcio commented:
Have you already joined our discord?

sobczal2 commented:
> Have you already joined our discord?

Yes :)
